tests/core/asyncio/test_asyncio_subscriptions_api.py

```diff
@@ -0,0 +1,223 @@
+    assert StreamEvent in remote.subscribed_messages
+    assert StreamEvent in subscriber.subscribed_events
+
+    # Broadcast two of the inherited events and then the correct event.
+    await other.broadcast(InheretedStreamEvent())
+    await other.broadcast(InheretedStreamEvent())
+    await other.broadcast(StreamEvent())
+
+    # wait for a received event, finishing the stream and
+    # consequently the subscription
+    event = await task
+    assert isinstance(event, StreamEvent)
+
+
+class SubscribeEvent(BaseEvent):
+    pass
+
+
+@pytest.mark.asyncio
+async def test_asyncio_subscribe_updates_subscriptions(pair_of_endpoints):
+    subscriber, other = pair_of_endpoints
+    remote = other._outbound_connections[subscriber.name]
+
+    assert SubscribeEvent not in remote.subscribed_messages
+    assert SubscribeEvent not in subscriber.subscribed_events
+
+    received_events = []
+
+    # trigger a `wait_for` call to run in the background and give it a moment
+    # to spin up.
+    subscription = await subscriber.subscribe(SubscribeEvent, received_events.append)
+    await asyncio.sleep(0.01)
+
+    # Now that we are within the wait_for, verify that the subscription is active
+    # on the remote
+    assert SubscribeEvent in remote.subscribed_messages
+    assert SubscribeEvent in subscriber.subscribed_events
+
+    # Broadcast and receive the second event, finishing the stream and
+    # consequently the subscription
+    await other.broadcast(SubscribeEvent())
+    # give time for propagation
+    await asyncio.sleep(0.01)
+    assert len(received_events) == 1
+    event = received_events[0]
+    assert isinstance(event, SubscribeEvent)
+
+    # Ensure the event is still in the subscriptions.
+    assert SubscribeEvent in remote.subscribed_messages
+    assert SubscribeEvent in subscriber.subscribed_events
+
+    subscription.unsubscribe()
+    # give the subscription removal time to propagate.
+    await asyncio.sleep(0.01)
+
+    # Ensure the event is no longer in the subscriptions.
+    assert SubscribeEvent not in remote.subscribed_messages
+    assert SubscribeEvent not in subscriber.subscribed_events
+
+
+@pytest.fixture
+async def client_with_three_connections(ipc_base_path):
+    config_a = ConnectionConfig.from_name("server-a", base_path=ipc_base_path)
+    config_b = ConnectionConfig.from_name("server-b", base_path=ipc_base_path)
+    config_c = ConnectionConfig.from_name("server-c", base_path=ipc_base_path)
+
+    async with AsyncioEndpoint.serve(config_a) as server_a:
+        async with AsyncioEndpoint.serve(config_b) as server_b:
+            async with AsyncioEndpoint.serve(config_c) as server_c:
+                async with AsyncioEndpoint("client").run() as client:
+                    await client.connect_to_endpoint(config_a)
+                    await client.connect_to_endpoint(config_b)
+                    await client.connect_to_endpoint(config_c)
+
+                    yield client, server_a, server_b, server_c
```

Comment:
If this was not just a fixture, I'd be concerned about the rightwards drift and recommend to combine this into a single `async_with` statement.


```python
    async with AsyncioEndpoint.serve(
        config_a
    ), AsyncioEndpoint.serve(
        config_b
    ), AsyncioEndpoint.serve(
        config_c
    ), AsyncioEndpoint("client").run()) as server_a, server_b, server_c, client:
        yield client, server_a, server_b, server_c
```

But given this is just a fixture with a single `yield` statement I think this is fine :+1: